### ISSUE TITLE
static property inheritance is incorrect

### ISSUE THREAD

**Issue by blesh** [Microsoft/TypeScript#4628]
I've seen that you've already closed many issues on this. But I'm reporting another one in hopes to wear you down ;)

It seems I can't change the signature of a static function in a derived class. The fact that the subclass is at all aware of the static functions on the super class is REALLY, REALLY strange.

I developed in C# for years, so I know what you were *trying* to do... but given that you don't support a `new` keyword like C# does, I think this behavior is really wrong. If you're looking to target being a superset of ES6/ES7, you need to correct this behavior ASAP. Generally in JavaScript, static properties are not copied to subclasses unless the copy is explicit.

I ran into this issue with [RxJS Next](https://github.com/ReactiveX/RxJS), where I Subject must inherit from Observable, but they both need *different* static `create()` signatures. So I end up hacking around and fighting TypeScript.

**Comment by mhegazy** [comment]
The intuition we tried to match is that a specialized class can stand in for its base; which is a general intuition that an OOP developer would expect. 

can you expand more on your scenario, and assuming a base class is needed but not expected to be generic enough to handle the child?

**Comment by blesh** [comment]
The issue is specifically where [Observable.create](https://github.com/ReactiveX/RxJS/blob/e0d27ba7e818d772a81e9823b2f62ab28f71db86/src/Observable.ts#L27-L29) and [Subject.create](https://github.com/ReactiveX/RxJS/blob/e0d27ba7e818d772a81e9823b2f62ab28f71db86/src/Subject.ts#L25-L27) need to have different signatures.

As you can see, I've found a workaround, but it's not ideal. I'd rather not copy static methods between classes at all.

**Comment by blesh** [comment]
I guess ES6 does pull the static functions over. But I'm able to redefine them at will in ES6. TypeScript complains. It's problematic for functions that are idiomatic in many JavaScript libraries like static `create` functions.

**Comment by RyanCavanaugh** [comment]
Summary of discussion from the Slog today:
* Static inheritance *is* part of the ES6 spec, but...
* Some people use the static side of classes (i.e. constructor functions with `static` properties) polymorphically, and some don't
* The first group wants this check as much as they want general substitutability checks for the *instance* side of classes
* The second group doesn't care at all if their `static` members align or not
* The first group *should* have their substitutability failures detected at the use sites, which is worse than at declaration sites, but still (mostly) works
* The second group is just out of luck in our current design

Conclusion: Try removing the check for assignability of the static side of classes and see what kind of errors go away (mostly in the baselines, since we mostly have clean code in our real-world-code baselines). Proceed depending on those results.

The `Committed` tag here is tentative.

**Comment by SinoBoeckmann** [comment]
So... and now there a plans to change this? I mean, this is pretty old now? Plans for a survey or something?

**Comment by gunn4r** [comment]
We're trying to extend base `Promise` to create a cancellable promise class. We ran into this issue when we wanted to change the static methods to return cancellable promises instead of base promises. Not sure of a good work around to the issue. For now I think we're going to implement those static method under different names (`resolveAsCancellable` for example).

**Comment by benwerner01** [comment]
Static factory methods are a useful design paradigm, for instance when object creation relies on data that needs to be asynchronously fetched a static `create()` method can encapsulate the data fetching, and return the object type (as such fetches cannot be conducted directly in the constructor). Requiring that in inheritance each static method with the same name must have compatible types restricts this significantly (for `create()`, this would enforce the use of the same parameters).

**Comment by marijnh** [comment]
I agree that this should at least be a strictness option that can be turned off. In some circumstances people might actually use constructor values and their static methods polymorphically, but I don't think that's a widely understood expectation, and this restriction gets in the way of entirely reasonable code such as static constructor methods with the same name but different parameter types.

**Comment by Sayan751** [comment]
Recently ran into this as well.

```ts
type Class<T> = {
    readonly prototype: T;
    new(...args: any[]): T;
};

class Foo {
    public static create<TFoo extends Foo = Foo>(this: Class<TFoo>, model: Partial<TFoo>): TFoo {
        return new this(model.propertyA);
    }

    public constructor(public propertyA?: string) { }
}

class Bar extends Foo {

    public static create<TBar extends Bar = Bar>(this: Class<TBar>,model: Partial<TBar>): TBar {
        return new this(
            model.propertyB,
            model.propertyA,
        );
    }

    public constructor(
        public propertyB: number,
        public propertyA?: string,
    ) {
        super(propertyA);
    }
}
```

Intuitively speaking, as `Bar#create` breaks the prototypal "link", I find this restriction bit of an annoyance. Refer this [playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwgNgQwM5IDwBUB8UC8UDeAUFCVAE4QIAmA9gHZwhRhk3BvgQBcU6A3MVJ0IAdwAUAOikIyAcyQ8EdEAG0AugEoe-QgF8BhAMaIUUAGI0aBQSTABXAEZwAloahJgCYK6iGKXiAwLKwgAD2AIOiokc0tcWJpMMWAAC2cFWBM0dGDMABooAFsaKgg4HgAFGW8EOCDLTC1eYOtSNvIIYDsyOihhEShU9LEbdrbi0rgJFhpIMlAAQTzRto0BNt1CUfsnH0N6DzI7Q3YyEbHbRxc3GbnFgH4eQ+c6WWXVgihNzaMsqAAhGRQMIRKIxFpEbZXHweLx7fwRDCAsjA8KRaIAoF4ZGYFakc4XNpDDLwZDZHHvQmkCZlSrVZy1JEyXGEproZGtC4ULo9PqiQZpJAEqlFEplaasO4gf6UkU0qa3CDzEBLPEkNajH5tHbXXwHYBHE40M5q5jQm6SpWgf48Oh2QoOJWy9o6nyK5ULR7uA0vN6jDScsZIOxzYVU92LZ2rdakb5AA). 

Contextually, I also tried to use the `InstanceType` utility class to find a workaround. But w/o much luck. Here is the [playground link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ0PQA4FcBGIAlsNBAC5hnHTABOAppfQDwAqCS9AHmfQHYATGGQCeWevABmcRNAC80UeKkz4APgAUZABaEIALmjtEAGmgBbeAPohDABTC0qYEMwCSfcmD7B6rMSzG6moAlIYeXj5+AWwcaijomEkMZDi0fNB89ADuiroQGolJxRZWNgB0WLTw4k4iAIImRcUh0JDQERRR-uKxiGoA3M0AvqjNuATUwPCeZLQ4wGTwtIUlGBNEJFU19HX1APyG5LSEfADmTUmtyNCjo6igkDAAQo7Q3LyCMBwJ4-ibpAoVBIdEYvDYr1o7x4-CEigCKkh8nhymkkLUzSSqzWSR0ekMrEhlxxmEs1ls0AcTkILncs28vh6gXRGJxYQ69O6MUJjniaBxKTSGSyuTxBUxazJFW2tVEzwAhMSSaVyZVqrKGkqWm0YJ0GdFejzaIMRmMkhspjNjgslisJet-tQZbs5YY+DhzHhdlrzY6tuqXQ1DoCTuctdd7aQcLVscrnXsfZgQkMkncxtNZtA8G8FJDyqCmBpkPHRPVDAAiACM5bMJZEz0MleGyYeVvgIHo5RA8DOGgAUgBlADyADlysdTmdCJIRBps7QQmZ59BTpFfIjHEu3quuuvpBxk0A) for that.

**Comment by shicks** [comment]
This is causing a number of problems trying to migrate Closure Library to TypeScript.  In particular, many of our legacy classes make common use of the pattern

```typescript
class Component {}
namespace Component {
  export enum EventType { FOO }
}

class Button extends Component {}
namespace Button {
  export enum EventType { BAR }
}
```

But this [complains](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDCD2BbADvAdgUzQF2gbwF8AoNMRDCZMYDOJVTHfI6aDAD1QCdcsBXRNACiANyzYAKgE9ktPNABiAeSXRixIqEgwAQn2zZ0bdtiwATGAhTpx+YqXKVqtPQaN4Wx7rzQDhYnGlZfGgdAEEAJTUiAiA) about class-side inheritance since the two `EventType` enums are incompatible.

In actuality, subclass constructors are not generally substitutable for one another, so enforcing it at the class level seems misguided.  In particular, all the existing enforcement [does nothing to prevent the following error](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDCD2BbADvAdgUzQF2gbwChppkAnASwDcxsNpSMwATdEAT2gA9oBeaARgDcRaMHQRspAK7Bs8UgAoGzVhzEp0WbABU2yDAC5oEimgDmASnwBfAiIk1ywUctoKLRhBsw58I4gzYUqRo0JgA7tDYABbkEAoA5GZYGBTACRbCxLa2BKCQMABCUthyoRictGhMMF6oPriExGRUNHTKLGjs0Bx8AExZouKSMnKKHarG5ABehmFSiABGqVZNxMZS+ooJiyVlGYPE5ABm0ArYehjwpxAzdACEPHwJaAvLpBlR0aTwkQCipB+23g8GQEHuBxEOTsxVK6AAdMBXBh3IIgA):

```
class Component {
  private readonly x = 1;
  constructor(readonly componentType: string) {}

  static create(): Component {
    return new this('generic');
  }
}

class Button extends Component {
  private readonly y = 2;
  constructor(readonly size: number) {
    super('button');
    if (typeof size !== 'number') throw Error('oops!');
  }
}

Button.create();
```

The correct solution would be to to ban unannotated use of `this` in static methods of non-final classes and require explicit specification:

```
class Component {
  static create<T extends typeof Component>(this: T) { ... }
}
```

This triggers an error at the call site of `Button.create()` since TypeScript can already tell that `Button` is not assignable to `Component`.  Given this, it makes sense to loosen up on the requirement that overridden static properties must be compatible.  TypeScript is smart enough to know when it matters.

**Comment by realh** [comment]
The trouble is that inheritance of static methods works differently from inheritance of instance methods. Here is a simplified example based on a real world problem I've encountered. GObject has this static method:

```
static install_property(property_id: number, pspec: ParamSpec): void
```

GtkSettings is derived from GObject and has this static method:

```
static install_property(pspec: ParamSpec): void
```

If they were instance methods, yes that would cause problems whether you were using JS or TS, but as they're static methods you're always going to qualify them when you call them eg `GtkSettings.install_property(pspec)`, so there shouldn't be any confusion over which method gets invoked. I know there can be problems when using `this` in static methods, but that isn't an issue here because these are bindings for native code.

However, the only way I can declare Settings' method in Typescript is by tricking it into thinking it's an overloaded method ie add declarations for both signatures to Gtk.Settings. This is the exact opposite of what Typescript is supposed to do in terms of type safety etc!

**Comment by Luxcium** [comment]
[Class static side 'typeof Maybelist' incorrectly extends base class static side 'typeof Monad'.](https://github.com/Luxcium/maybelist/blob/86ae02eda6eb4738cb215c01dc42c7a65d00ce84/src/core/maybelist/maybelist.ts#L7)

```log
Class static side 'typeof Maybelist' incorrectly extends base class static side 'typeof Monad'.
  Types of property 'fromValueOf' are incompatible.
    Type '<xTVal>(value: Functor<xTVal[]>) => Maybelist<xTVal, xTVal[]>' is not assignable to type '<yTVal>(value: Functor<yTVal>) => Monad<yTVal>'.
      Types of parameters 'value' and 'value' are incompatible.
        Type 'Functor<yTVal>' is not assignable to type 'Functor<unknown[]>'.
          Type 'yTVal' is not assignable to type 'unknown[]'.ts(2417)
```

[![Screenshot_src_classes_monad](https://raw.githubusercontent.com/Luxcium/maybelist/next/from-20201112-0.0.0/Screenshot_src_classes_monad_monad.ts%20-%20maybelist_20201218_022806.png)](https://github.com/Luxcium/maybelist/blob/86ae02eda6eb4738cb215c01dc42c7a65d00ce84/src/core/maybelist/maybelist.ts#L9)


[public static fromValueOf in class Maybelist uses generic xTVal line 24](https://github.com/Luxcium/maybelist/blob/86ae02eda6eb4738cb215c01dc42c7a65d00ce84/src/core/maybelist/maybelist.ts#L24)

```typescript
// in class Maybelist<Val = unknown, MLVal extends Array<Val> = Val[ ]> extends Monad<MLVal> :
  public static fromValueOf<xTVal>(value: Functor<xTVal[]>): Maybelist<xTVal> {
    return Maybelist.of<xTVal>(
      (Monad.from<xTVal[]>(value.clone).fork as unknown) as xTVal,
    );
  }
```
[public static fromValueOf in class Monad uses generic yTVal  line 15](https://github.com/Luxcium/maybelist/blob/86ae02eda6eb4738cb215c01dc42c7a65d00ce84/src/classes/monad/monad.ts#L15)
```typescript
// in class Monad<MVal = unknown> extends Chain<MVal> implements IMonad<MVal> :
   public static fromValueOf<yTVal>(value: Functor<yTVal>): Monad<yTVal> {
    return Monad.from<yTVal>(value.clone);
  }
```

**Comment by justinfagnani** [comment]
@realh: "If they were instance methods, yes that would cause problems whether you were using JS or TS, but as they're static methods you're always going to qualify them when you call them eg `GtkSettings.install_property(pspec)`"

This is not always true. My team has a lot of code where static members are the implementation of a static interface that is accessed generically on constructor instances.

This is a simple example of configuration statics that cause problems because static declarations are allowed to narrow the static type of a supertype. Once the type is narrowed, a further subclass can't widen it again without a cast:

```ts
class A {
  static p: string | string[];
}

class B extends A {
  static p = 'B';
}

// Error
// Class static side 'typeof C' incorrectly extends base class static side 'typeof B'.
//  Types of property 'p' are incompatible.
//    Type 'string[]' is not assignable to type 'string'.(2417)
class C extends B {
  static p = ['B'];
}

function getP(ctor: typeof A): string {
  return typeof ctor.p === 'string' ? ctor.p : ctor.p.join(',');
}
```

What I think I want here is that without an explicit type annotation, `B.p` kept the type of B's superclass. I'm admittedly unsure what other problems that might cause.

**Comment by shicks** [comment]
I'd argue that the extends-time enforcement is the fundamental problem.  The check ultimately needs to be at the usage site.  In your example, the call to `getP(C)` will typecheck just fine because `typeof C` is a subtype of `typeof A`, but this is _not_ just because `C` is a subtype of `A` (subtypeness is not invariant across `typeof`, though that's a partial requirement): it's also necessary that the static side be completely compatible.  _That's_ when the type checker should enforce the static side types.  And if `D extends A` is never passed as `typeof A` then it can define `p` to be whatever type it wants.  (For subtyping compatibility, it's also necessary that the constructor arguments be compatible; if that's not needed for some application, a structural type or interface like `{p: string|string[]}` could be more broadly useful.

**Comment by jerrys-msft** [comment]
Yeah, this seems fatally broken to me, especially since it applies to _private_ statics as well. And I can see no advantage to the current behavior that could possibly outweigh the loss of the factory method pattern.

**Comment by shicks** [comment]
There's also unfortunately no remotely reasonable way to work around this.  Any suppression has to happen at the level of the entire class, which loses a lot of type checking.

**Comment by Arturokin** [comment]
Any update on this? i'm stuck, i want to have different parameters on parent static function and in child static function, i don't see necessary that Typescript does this with static functions

**Comment by cataclym** [comment]
Please, we need an update on this

**Comment by richardgirges** [comment]
At the very least, is there any chance we can get the error to appear at the static method level, as opposed to the class? This would allow us to suppress it with `// @ts-expect-error` without having to disable errors on the entire class.

**Comment by benlesh** [comment]
FWIW: I'm going to stop watching this, but I'll leave it open for others. I'm okay with the current behavior because I try to avoid inheritance like the plague now. ... and static methods, TBH. haha. Partially because of this issue.

**Comment by shicks** [comment]
I still believe this is worth fixing (see my earlier comments), but I've come up with a free-after-typechecking workaround that at least unblocks me, so I thought I'd share it here:

```typescript
type OmitStatics<T, S extends string> =
    T extends {new(...args: infer A): infer R} ?
        {new(...args: A): R}&Omit<T, S> :
        Omit<T, S>;
```

Example usage:
```typescript
class Base {}
namespace Base {
  export enum EventType { A }
}
class Sub extends (Base as OmitStatics<typeof Base, 'EventType'>) {}
namespace Sub {
  export enum EventType { B }
}
```

TypeScript fully understands the correct class relationship, but it doesn't complain about the otherwise-incompatible override anymore.

Warning: nothing prevents you from calling an incompatibly-overridden static method unsafely (e.g. `Base`'s `static foo()` calls `this.bar(42)` but `Sub` incompatibly overrides `static bar(arg: string)`).  If you want that extra protection, you could take a broader approach by eliminating _all_ statics from the base class (this has the added "benefit" of enforcing that statics are only called directly on the classes they're defined on, which I'd consider best practice as long as you're not relying on the class-side dynamism):

```typescript
type OmitAllStatics<T extends {new(...args: any[]): any, prototype: any}> =
    T extends {new(...args: infer A): infer R, prototype: infer P} ?
        {new(...args: A): R, prototype: P} :
        never;
class Sub extends (Base as OmitAllStatics<typeof Base>) { /* ... */ }
```

[Playground examples](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gtgS2AZWAQ2AgxgZwDwAqANFMlBAB7AQB2AJjlDsAE4I0DmAfFALwBQUIVALkqtBlADeNCAHcAFADoVaFhxwAuKOwBmEFlACCASm16DUAEoBfKAH5Bw5zPnLV6rcbPWbAMngkQhJkHk0nZyFA4GDSLgBufn5QSFhEYCMAG0zUDGx8UUpqekZXRRUlNQ1tNBoQAG0AXR9akBIwFgB7YG7wCBq6mx4BZ0LxEulZco9qnRp9Q1Nzecsrdq6elP65hagABTtHSOEy90rPbSXrde7eyG0DqG1ZADcDRP4sTLQcRgAhH7QKRODoIF4YaAUACEzwgbxYiRs-BoaDgEBwYDQWGgAJwQKclDAnRYwHINAArnAoAAxTqdaTGEh-KBIpGfTo0ZikckAIwAJMhyZAWF8fl5orlMLg8FtOrooLiICQAOS0zrK4YKwGJUW-bk8sTFSSC-mC4W60ogtjg6hQEAwqCvd78JEotEYrHQE3SAkUIkksmUml0hlGEgAYRZLqSFv1ACZDRJGApFVAfmkkFkcugpfhZfLFVwTD6hKCbdAAF4Op0I6Nu9GY7Hxktif2k2hBtWhkgAESjSKAA)

**Comment by zimtsui** [comment]
In functional type theory, a type is defined as a set of object along with a set of operation. It's analogous to an algebraic structure except it doesn't have to comply with closure axiom.

The operations above may have the type as both its input or output.

However, in TypeScript, method call is written as `a.f()` rather than `f(a)`. This syntax cannot describe those operations which have another 3-party type as their input.

For example, there is a type A, which have 2 operations:

- capture an object's state as a snapshot.
- restore as an object from a snapshot.

```haskell
capture :: A -> Snapshot
restore :: Snapshot -> A
```

In TypeScript, you can capture like `a.capture()`. But the `restore` operation's input is not A, you cannot write somthing like `XXX.restore()`, so you have to write it as a static method:

```ts
class A {
	public capture(): Snapshot;
	public static restore(s: Snapshot): A;
}
const a = A.restore(snapshot);
```

Therefore, the semantic meaning of static methods of a class is the algebra-like operations as well.

This is why many non-functional OOP language all deicide to inherit static members automatically.

**Comment by steveetm** [comment]
Is the following the same or a bug?
```
interface BaseEntity {
    id: number;
}

interface ExtendedEntity extends BaseEntity {
     someextendedprop: number;
}

class Base<T extends BaseEntity> {
        public static async bulkCreate<E extends BaseEntity, M extends Base<E>>(
            this: { new(): M },
        ): Promise<M[]> {
            return [new this];
        }
        doSomething():T {
            return {} as T;
        }
};


class Sub<T extends ExtendedEntity> extends Base<T> {
    public static async bulkCreate<E extends ExtendedEntity,M extends Sub<E>>(
        this: { new(): M },
    ): Promise<M[]> {
        return [new this];
    }
};
```

Strange, as if I remove `doSomething` tha static complain on class disappears. But I don't really understand what is the problem, the subclass returns instances which are compatible with base class instances.
* If I have only the static declaration, everything **is ok.** 
* If I have only the member functions (as generic), everything **is ok.** 
* If I have both : 
```
Class static side 'typeof Sub' incorrectly extends base class static side 'typeof Base'.
  Types of property 'bulkCreate' are incompatible.
    Type '<E extends ExtendedEntity, M extends Sub<E>>(this: new () => M) => Promise<M[]>' is not assignable to type '<E extends BaseEntity, M extends Base<E>>(this: new () => M) => Promise<M[]>'.
      The 'this' types of each signature are incompatible.
        Type 'new () => M' is not assignable to type 'new () => Sub<ExtendedEntity>'.
          Type 'M' is not assignable to type 'Sub<ExtendedEntity>'.
            Type 'Base<E>' is not assignable to type 'Sub<ExtendedEntity>'.
              The types returned by 'doSomething()' are incompatible between these types.
                Type 'E' is not assignable to type 'ExtendedEntity'.
                  Property 'someextendedprop' is missing in type 'BaseEntity' but required in type 'ExtendedEntity'.(2417)
```

**Comment by shicks** [comment]
Yes, this is the same issue.  The declared type of `Sub.bulkCreate` is _not_ a subtype of the declared type of `Base.bulkCreate`.  TypeScript is expecting the static methods to follow Liskov substitution, so that you could write

```
Sub.bulkCreate.call(Base);
```

But as you've written your `Sub.bulkCreate` signature, you've constrained the _input_ type `M` to be _narrower_ in the subclass than in the base class (it requires `this` to be `Sub<ExtendedEntity>`, rather than a `Base<BaseEntity>`).  This is somewhat confusing for two reasons: (1) because function parameters are contravariant, so `(arg: number) => void` is actually a _supertype_ of `(arg: unknown) => void`, even though `number` is a _subtype_ of `unknown`.  But also (2) TypeScript plays a little fast and loose with this rule by allowing _methods_ (i.e. using method shorthand rather than a property with a colon) to break this rule, so a lot of people get away with breaking it without realizing what they're doing.

The reason the error goes away when you remove `doSomething` is because TypeScript types classes structurally: so without that method the template parameter no longer shows up _anywhere_ in the type, so that `Sub<A> === Base<B>` for all `A` and `B`, as far as the type checker is concerned (and `Sub` doesn't add any other properties, either).  Removing the extra property from `ExtendedEntity` would also fix the error, since as you've written it `Sub<A> === Base<A>` already.

Back to the original issue, there should be nothing wrong with what you've written.  The type checker can see that `typeof Sub` is not assignable to `typeof Base`, and can enforce this just fine.  If you wanted to make it assignable, you can, but it should let you choose to have it not be assignable (which it's not).  You can use the workaround I posted earlier to remove `bulkCreate` from `Base` in the `extends` clause, but we shouldn't have to do that.

**Comment by steveetm** [comment]
@shicks reading your workaround again after this explanation just really made it super clear, I can't upvote this enough. I switch would be nice to to disable this check(or make it appear on the function and just ignore it there), but I really think this could work well for more complex cases we use. Thanks!

**Comment by hazae41** [comment]
How about a new keyword `override` in the child class, that would explicitly say the compiler to fuck off?

**Comment by Gnuxie** [comment]
So from what I can gather, the reason this issue is dead in the water is because of these [concerns](https://github.com/microsoft/TypeScript/issues/4628#issuecomment-149333638).

However, the inaction has come at the cost that you cannot share the name of your static factory methods. As I understand this exact same restriction isn't applied to constructors, as the issue would appear far more frequently. Which makes sense because it would make using constructors completely intolerable. Well this is an intolerable issue and you simply cannot use static factory methods that will clash. The type gymnast workarounds also all suck, and no one should reasonably incorporate them into their code. 

The workaround I wanted to use is probably something like this: pay for the pain in verbose names, use the class name again in your static factory methods such that `FooBar.factory` = `FooBar.FooBarFactory`. However, even this sucks. If I wanted the factory to have a short name, for example having a `Result` static factory on an `Error` type to wrap them in a `Result` type, then you're all out of luck. There are for sure other valid examples where shorter and consistent names are desired. 

What makes it acceptable for this issue to be parked? The hope that very few users come across it without quickly finding a suitable workaround, but there does seem to be a lot of confidence in that.

**Comment by christianschmitz** [comment]
> FWIW: I'm going to stop watching this, but I'll leave it open for others. I'm okay with the current behavior because I try to avoid inheritance like the plague now. ... and static methods, TBH. haha. Partially because of this issue.

Good point: "prefer composition over inheritance...". Hitting this issue in Typescript can help you decide in favor of composition.
